### ISSUE TITLE
After adding tcp ingress config, ingress controller going down causing internet-facing URLs inacccessible

### ISSUE THREAD

**Issue by rajeshgheware** [Kong/kubernetes-ingress-controller#1154]
### Summary

After adding tcp ingress config, it is going down frequently

**Kong Ingress controller version**  2.2  (kong image)

**Kong or Kong Enterprise version** : Kong

### Kubernetes version

**paste `kubectl version` output**  Server v1.19.7

### Environment

- **Cloud provider or hardware configuration**: AWS 
- **OS** (e.g. from /etc/os-release): Ubuntu 20.04.1 LTS
- **Kernel** (e.g. `uname -a`): 5.4.0-1035-aws
- **Install tools**: KOPS
- **Others**: docker://19.3.14

### What happened
After adding TCP ingress config KONG_STREAM_LISTEN, ingress controller going down every few hours resulting all our internet-facing URL inaccessible

** Error observed (kubectl logs -f -n kong deploy/ingress-kong proxy) **
```
2021/04/02 07:10:31 [error] 25#0: *12612660 lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/init.lua:634: body must be a nil or a string
stack traceback:
coroutine 0:
	[C]: in function 'error'
	/usr/local/share/lua/5.1/kong/pdk/response.lua:849: in function 'after'
	/usr/local/share/lua/5.1/kong/init.lua:634: in function 'preread'
	preread_by_lua(nginx-kong-stream.conf:92):2: in main chunk while prereading client data, client: 100.102.0.0, server: 0.0.0.0:20130
```
**TCP config patch** 
```
kubectl patch deploy -n kong ingress-kong --patch '{
  "spec": {
    "template": {
      "spec": {
        "containers": [
          {
            "name": "proxy",
            "env": [
              {
                "name": "KONG_STREAM_LISTEN",
                "value": "0.0.0.0:20130"
              }
            ],
            "ports": [
              {
                "containerPort": 20130,
                "name": "stream20130",
                "protocol": "TCP"
              }
            ]
          }
        ]
      }
    }
  }
}' 
```
```
kubectl patch service -n kong kong-proxy --patch '{
  "spec": {
    "ports": [
      {
        "name": "stream20130",
        "port": 20130,
        "protocol": "TCP",
        "targetPort": 20130
      }
    ]
  }
}'
```
**TCP Ingress**
```
apiVersion: configuration.konghq.com/v1beta1
kind: TCPIngress
metadata:
  name: tcp-ingress
  namespace: prod
  annotations:
    kubernetes.io/ingress.class: kong
spec:
  rules:
  - port: 20130
    backend:
      serviceName: server
      servicePort: 20130
```
### Expected behavior
Ingress controller should be up all the time without needing to restart the deployment every few  hours

### Steps To Reproduce

1. Create kong kubernetes ingress: kubectl create -f https://bit.ly/k4k8s
2. Add TCP Ingress config (see the what happened section)

### Workaround
Whenever an alert email is received, the team executes below
kubectl rollout restart deploy ingress-kong -n kong

**Comment by rajeshgheware** [comment]
Another error spotted in kong pods is:

2021/04/02 08:01:38 [error] 27#0: *479908 connect() failed (113: Host is unreachable) while connecting to upstream, client: 100.120.0.0, server: kong, request: "GET /<resource name> HTTP/2.0", upstream: "http://100.124.0.2:80/<resource name>", host: "<prod domain>", referrer: "https://<prod URL>"

However HOST pod is running smoothly

**Comment by mflendrich** [comment]
@rajeshgheware any plugins enabled?

**Comment by rajeshgheware** [comment]
kind: KongClusterPlugin
metadata:
  name: prometheus
  annotations:
    kubernetes.io/ingress.class: kong
  labels:
    global: "true"
plugin: prometheus

**Comment by rajeshgheware** [comment]
@mflendrich  any update?

**Comment by rainest** [comment]
The "body must be a nil or a string" error appears to be due to improper error handler code (https://github.com/Kong/kong/issues/7073#issuecomment-827123798) in the event of a balancer failure (essentially, some sort of issue that prevents Kong from selecting an upstream host). I'm not immediately sure what a proper response would look like there (TCP doesn't have a mechanism for reporting such errors), though we can probably log something informative. That's probably beside the point, since ultimately balancer resolution failed anyway.

The "113: Host is unreachable" error is an [ICMP unreachable error](https://en.wikipedia.org/wiki/Internet_Control_Message_Protocol#Destination_unreachable); there was some sort of layer 3/network issue at that time (e.g. there was no route to that host). The upstream host may be running, but the kernel wasn't able to find a path to it.

Are the upstream addresses you see in the unreachable errors indeed a current address for some upstream? It could be that there was some issue updating the endpoint list, though _all_ upstreams becoming unreachable suggests a lower-level issue.

If the addresses are current, and you leave the instances online when this occurs, what do you see if you exec into the Kong container and attempt to traceroute to an upstream address? That may provide some indication as to what's failing in routing. Do your container daemon logs on the kubelet indicate any problems with the virtual interfaces used for the Kong container?

**Comment by rainest** [comment]
Upstream has merged a fix: https://github.com/Kong/kong/issues/7073#issuecomment-831259042

The original issue mentioned will be fixed in Kong 2.4.2 (not yet released), and we know of no controller-level issue that'd persistently keep unavailable endpoints around (for the second issue mentioned)/have no similar reports of that. Will go ahead and close this out.
